src/parser/nmodl_driver.cpp

```diff
@@ -24,60 +24,132 @@ std::shared_ptr<ast::Program> NmodlDriver::parse_stream(std::istream& in) {
     NmodlLexer scanner(*this, &in);
     NmodlParser parser(scanner, *this);
 
-    this->lexer = &scanner;
-    this->parser = &parser;
-
     scanner.set_debug(trace_scanner);
     parser.set_debug_level(trace_parser);
     parser.parse();
     return astRoot;
 }
 
-//// parse nmodl file
-std::shared_ptr<ast::Program> NmodlDriver::parse_file(const std::string& filename) {
+std::shared_ptr<ast::Program> NmodlDriver::parse_file(const std::string& filename,
+                                                      const location* loc) {
     std::ifstream in(filename.c_str());
-    stream_name = filename;
-
     if (!in.good()) {
-        logger->error("Can not open file : {}", filename);
-        return nullptr;
+        std::ostringstream oss;
+        if (loc == nullptr) {
+            oss << "NMODL Parser Error : ";
+        }
+        oss << "can not open file : " << filename;
+        if (loc != nullptr) {
+            parse_error(*loc, oss.str());
+        } else {
+            throw std::runtime_error(oss.str());
+        }
+    }
+
+    auto current_stream_name = stream_name;
+    stream_name = filename;
+    auto absolute_path = utils::cwd() + utils::pathsep + filename;
```

Comment:
A quick question - `absolute_path` here is the dir where nmodl is launched or the directory of input mod file? In case of MOD file usage in scientific projects, IIRC,  `INCLUDE` file paths are related to the `.mod`  file that is including the file. For example, if `/a/b/c/x.mod` includes `INCLUDE "y.inc"` then `y.inc` is in `/a/b/c` directory.

Can you check if it's the same assumption here?

Reply:
The specification of the INCLUDE directive is different from what you say
https://neuronsimulator.github.io/nrn/py_doc/modelspec/programmatic/mechanisms/nmodl.html#include

> If the full path to the file is not given, the file is first looked for in the current working directory, then in the directory where the original .mod file was located, and then ...

And this behavior is implemented in `NmodlDriver::parse_include` member method.

Reply:
👍